### PR TITLE
Fixup for `JSON::Serializable` on certain recursively defined types

### LINKED ISSUE #13429: Another recursive json bug, after crystal 1.8
failed with 1.8.1 (https://play.crystal-lang.org/#/r/ezhy), but works with 1.7.3 (https://play.crystal-lang.org/#/r/ezhx)

```crystal
require "json"

abstract class Base
  include JSON::Serializable

  use_json_discriminator("type", {"a" => A, "b" => B, "c" => C})
end

class A < Base
end

class B < Base
  property source : Base
  property value : Int32 = 1  # <= BUG is here, if i remove `= 1`, it compiles
end

class C < B
end

p Base.from_json %q({"type": "c", "source": {"type": "a"}, "value": 1})
```

```
Showing last frame. Use --error-trace for full trace.

In /home/kostya/Downloads/crystal-1.8.1-1/share/crystal/src/json/serialization.cr:181:9

 181 | def initialize(*, __pull_for_json_serializable pull : ::JSON::PullParser)
           ^---------
Error: instance variable '@value' of B+ was not initialized in this 'initialize', rendering it nilable

```

### PR DESCRIPTION
**Pull request by kostya**
i don't know, correct this or not, but if fixes, and all specs seems pass. fixes: #13429

### CODE REVIEW

**Comment by straight-shoota** [comment]
This essentially reverts part of #13344 which apparently wasn't necessary? I suppose the relevant part of that PR was casting the type of the default nil value?

It's not exactly clear why this fixes things. The compiler is doing some weird stuff here.

**Comment by kostya** [comment]
yes looks like compiler bug when using `typeof`. I will add for yaml too, if its ok.

File: spec/std/json/serializable_spec.cr [hunk 1]
```diff
@@ -471,6 +471,25 @@ class JSONSomething
...
 
+module JsonDiscriminatorBug
+  abstract class Base
+    include JSON::Serializable
+
+    use_json_discriminator("type", {"a" => A, "b" => B, "c" => C})
+  end
+
+  class A < Base
+  end
```
Comment (HertzDevil) [inline spec/std/json/serializable_spec.cr:482]: `A` is not necessary for a minimal reproduction

File: src/json/serialization.cr [hunk 2]
```diff
@@ -202,9 +203,9 @@ module JSON
         # recursively defined serializable types
         {% for name, value in properties %}
           %var{name} = {% if value[:has_default] || value[:nilable] %}
-                         nil.as(::Nil | typeof(@{{name}}))
+                         nil.as(::Nil | Union({{value[:type]}}))
```
Comment (HertzDevil) [inline src/json/serialization.cr:206]:
```suggestion
                         nil.as(Union(::Nil, typeof(@{{name}})))
```

For the others simply wrap the `typeof` alone inside `Union`

**Review by HertzDevil — CHANGES_REQUESTED** [review]
My guess is `typeof(@source).new(pull)` somehow instantiates `Base+.new` which breaks things, while `Union(typeof(@source))` devirtualizes the metaclass type.

YAML would be nice.

File: spec/std/json/serializable_spec.cr (hunk 1 above)
Comment (kostya) [reply spec/std/json/serializable_spec.cr:482]: A need because it leaf node, without it would infinite recursion.

**Review by straight-shoota — APPROVED** [review]

**Review by HertzDevil — APPROVED** [review]
Fixes #13429

**Comment by oprypin** [comment]
I am also observing some JSON bug, possibly slightly different, in Crystal 1.8.1.
https://github.com/oprypin/crystal-imgui/actions/runs/4771237772/jobs/8483033672
And I confirm that this PR fixes it.

**Pull request merged**